File: examples/laya/http_provider.py
"""Bounded Laya HTTP producer; refuses redirects and plaintext remote credentials."""
import http.client
import json
import os
from urllib.parse import urlsplit


def predict(request):
    base = urlsplit(os.environ.get('LAYA_URL', 'http://127.0.0.1:8000'))
    if (base.scheme not in ('http', 'https') or not base.hostname or base.username or base.password
            or base.path not in ('', '/') or base.query or base.fragment
            or (base.scheme == 'http' and base.hostname not in ('127.0.0.1', 'localhost', '::1'))):
        raise ValueError('LAYA_URL must be an HTTPS origin or loopback HTTP origin')
    headers = {'Content-Type': 'application/json'}
    key = os.environ.get('LAYA_API_KEY')
    if key:
        if any(ord(c) <= 32 or ord(c) == 127 for c in key):
            raise ValueError('Invalid Laya API key format')
        headers['Authorization'] = 'Bearer ' + key
    cls = http.client.HTTPSConnection if base.scheme == 'https' else http.client.HTTPConnection
    connection = cls(base.hostname, base.port, timeout=30)
    try:
        connection.request('POST', '/v1/systemone', json.dumps(request, allow_nan=False).encode(), headers)
        reply = connection.getresponse()
        if reply.status != 200:
            raise RuntimeError('Laya returned HTTP %s; inspect your local inference service' % reply.status)
        data = reply.read(2 * 1024 * 1024 + 1)
        if len(data) > 2 * 1024 * 1024:
            raise ValueError('Laya response exceeds 2 MiB')
        return json.loads(data)
    finally:
        connection.close()

File: examples/laya/http_provider.mjs
// Laya runs outside ChronoDB. Never forward its key through a redirect.
export async function predict(request) {
  const url = new URL(process.env.LAYA_URL ?? 'http://127.0.0.1:8000');
  if (!['https:','http:'].includes(url.protocol) || url.username || url.password || url.search || url.hash || url.pathname !== '/'
    || (url.protocol === 'http:' && !['127.0.0.1','localhost','[::1]'].includes(url.hostname))) throw new Error('Use an HTTPS or loopback HTTP Laya origin');
  const headers = {'Content-Type':'application/json'};
  if (process.env.LAYA_API_KEY) {
    if (/[\s\x00-\x1f\x7f]/.test(process.env.LAYA_API_KEY)) throw new Error('Invalid Laya key');
    headers.Authorization = 'Bearer ' + process.env.LAYA_API_KEY;
  }
  const result = await fetch(new URL('/v1/systemone',url), {method:'POST',headers,body:JSON.stringify(request),redirect:'error',signal:AbortSignal.timeout(30000)});
  if (!result.ok) { await result.body?.cancel(); throw new Error(`Laya returned HTTP ${result.status}`); }
  const reader = result.body.getReader(), chunks=[];let total=0;
  try { for (;;) { const {value,done}=await reader.read();if(done)break;total+=value.length;
    if(total>2*1024*1024)throw new Error('Laya response exceeds 2 MiB');chunks.push(value);
  }} finally { await reader.cancel(); }
  return JSON.parse(Buffer.concat(chunks).toString('utf8'));
}
